File: Halachic Times/GeoNames/src/main/java/com/github/geonames/Toponym.java
/*
 * Copyright 2012, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.geonames;

import com.github.geonames.util.LocaleUtils;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

/**
 * GeoNames toponym POJO.
 *
 * @author Moshe Waisberg
 */
public class Toponym extends GeoNamesRecord {

    private String wikipediaURL;
    private final Map<String, AlternateName> alternateNamesMap = new HashMap<>();

    public Map<String, AlternateName> getAlternateNamesMap() {
        return alternateNamesMap;
    }

    public void setAlternateNames(Map<String, AlternateName> alternateNames) {
        this.alternateNamesMap.clear();
        if (alternateNames != null) {
            this.alternateNamesMap.putAll(alternateNames);
        }
    }

    public void setAlternateNames(Collection<AlternateName> alternateNames) {
        this.alternateNamesMap.clear();
        if (alternateNames != null) {
            for (AlternateName name : alternateNames) {
                this.alternateNamesMap.put(name.getName(), name);
            }
        }
    }

    public String getWikipediaURL() {
        return wikipediaURL;
    }

    public void setWikipediaURL(String wikipediaURL) {
        this.wikipediaURL = wikipediaURL;
    }

    public void putAlternateName(long geonameId, String language, String name) {
        putAlternateName(geonameId, language, name, false, false, false, false);
    }

    public void putAlternateName(long geonameId, String language, String name, boolean preferred, boolean shortName, boolean colloquial, boolean historic) {
        String languageCode = LocaleUtils.toLanguageCode(language);
        String languageCodeISO = LocaleUtils.getISOLanguage(languageCode);
        AlternateName alternateName = alternateNamesMap.get(languageCodeISO);
        if ((alternateName == null) || preferred) {
            alternateName = new AlternateName(language, name, preferred, shortName, colloquial, historic);
            alternateNamesMap.put(languageCodeISO, alternateName);
        } else if (!alternateName.isPreferred()) {
            if ((alternateName.isShortName() && !shortName)
                    || (alternateName.isColloquial() && !colloquial)
                    || (alternateName.isHistoric() && !historic)) {
                alternateName = new AlternateName(language, name, preferred, shortName, colloquial, historic);
                alternateNamesMap.put(languageCodeISO, alternateName);
            } else if (!historic && !colloquial && (languageCode.length() <= alternateName.getLanguage().length())) {
                alternateName = new AlternateName(language, name, preferred, shortName, colloquial, historic);
                alternateNamesMap.put(languageCodeISO, alternateName);
            } else {
                System.err.println("Duplicate name! id: " + geonameId + " language: " + language + " name: [" + name + "]");
            }
        } else {
            System.err.println("Duplicate name! id: " + geonameId + " language: " + language + " name: [" + name + "]");
        }
    }

    public String getBestName(String language) {
        String name = getName(language);
        return name != null ? name : getName();
    }

    public String getName(String language) {
        String languageCode = LocaleUtils.toLanguageCode(language);
        AlternateName alternateName = getAlternateNamesMap().get(languageCode);
        if (alternateName != null) {
            return alternateName.getName();
        }
        if (language == null) {
            return getName();
        }
        return null;
    }
}

File: Halachic Times/GeoNames/src/main/java/org/geonames/Toponym.kt
/*
 * Copyright 2012, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.geonames

import com.github.util.LocaleUtils.ISO3166_ISRAEL
import com.github.util.LocaleUtils.ISO3166_PALESTINE
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * GeoNames toponym.
 *
 * @author Moshe Waisberg
 */
@Serializable
open class Toponym {
    @SerialName("geonameId")
    var geoNameId: GeoNameId = 0

    @SerialName("toponymName")
    var toponymName: String? = null
        get() = field ?: name

    @SerialName("name")
    var name: String = ""

    /**
     * The ASCII name.
     */
    @SerialName("asciiName")
    var asciiName: String? = null

    @SerialName("alternateNames")
    var alternateNames: List<AlternateName>? = null

    @SerialName("continentCode")
    var continentCode: String? = null

    @SerialName("countryCode")
    var countryCode: String? = null
        set(value) {
            field = if (ISO3166_PALESTINE == value) {
                ISO3166_ISRAEL
            } else {
                value
            }
        }

    @SerialName("countryId")
    var countryId: GeoNameId? = null

    @SerialName("countryName")
    var countryName: String? = null

    @SerialName("population")
    var population: Long? = null

    @SerialName("elevation")
    var elevation: Int? = null
        get() = field ?: elevationSRTM ?: elevationAsterGDEM

    /**
     * The GDEM elevation.
     * ASTER Global Digital Elevation Model.
     */
    @SerialName("astergdem")
    var elevationAsterGDEM: Int? = null

    /**
     * The SRTM3 elevation.
     * Shuttle Radar Topography Mission.
     */
    @SerialName("srtm3")
    var elevationSRTM: Int? = null

    @SerialName("fcl")
    var featureClass: FeatureClass? = null

    @SerialName("fclName")
    var featureClassName: String? = null

    @SerialName("fcode")
    var featureCode: String? = null

    @SerialName("fcodeName")
    var featureCodeName: String? = null

    @SerialName("lat")
    var latitude: Double = 0.0

    @SerialName("lng")
    var longitude: Double = 0.0

    @SerialName("adminCode1")
    var adminCode1: String? = null

    @SerialName("adminId1")
    var adminId1: GeoNameId? = null

    @SerialName("adminName1")
    var adminName1: String? = null

    @SerialName("adminCode2")
    var adminCode2: String? = null

    @SerialName("adminId2")
    var adminId2: GeoNameId? = null

    @SerialName("adminName2")
    var adminName2: String? = null

    @SerialName("adminCode3")
    var adminCode3: String? = null

    @SerialName("adminId3")
    var adminId3: GeoNameId? = null

    @SerialName("adminName3")
    var adminName3: String? = null

    @SerialName("adminCode4")
    var adminCode4: String? = null

    @SerialName("adminId4")
    var adminId4: GeoNameId? = null

    @SerialName("adminName4")
    var adminName4: String? = null

    @SerialName("adminCode5")
    var adminCode5: String? = null

    @SerialName("adminId5")
    var adminId5: GeoNameId? = null

    @SerialName("adminName5")
    var adminName5: String? = null

    @SerialName("adminTypeName")
    var adminTypeName: String? = null

    @SerialName("timezone")
    var timeZone: TimeZone? = null

    @SerialName("bbox")
    var boundingBox: BoundingBox? = null

    @SerialName("wikipediaURL")
    var wikipediaURL: String? = null
}